### PR TITLE
fix: Bump objc2 to 0.5.0; bring icrate 0.1.0

### PR DESCRIPTION
**Pull request by DataTriny**
Replaces #200

- [x] https://github.com/madsmtm/objc2/issues/537,
- [x] https://github.com/rust-windowing/winit/pull/3256

- Update to latest objc2,
- Introduce icrate as a dependency,
- Remove handmade AppKit types in favor of auto-generated ones provided by icrate,
- Use type aliases from icrate as much as possible

The compile time of accesskit_macos increases a bit due to icrate. This shouldn't be a big issue for Rust users who will probably have it in their dependency chain already. Bindings generation will be slightly slower, but I don't think this should be a concern.

### CODE REVIEW

File: platforms/macos/src/node.rs [hunk 1]
```diff
@@ -327,22 +330,29 @@ struct BoxedData {
     node_id: NodeId,
 }
 
+struct PlatformNodeIvars {
+    boxed: Box<BoxedData>,
```
[the remark this answers is not in the document]
Comment (DataTriny) [reply platforms/macos/src/node.rs:334]: Thanks @madsmtm for the suggestion, I'll update accordingly.

File: platforms/macos/src/node.rs [hunk 2]
```diff
@@ -327,22 +330,29 @@ struct BoxedData {
...
     }
 
+    impl DeclaredClass for PlatformNode {
+        type Ivars = PlatformNodeIvars;
+    }
+
     unsafe impl PlatformNode {
-        #[sel(accessibilityParent)]
-        fn parent(&self) -> *mut NSObject {
+        #[method(accessibilityParent)]
```
[the remark this answers is not in the document]
Comment (DataTriny) [reply platforms/macos/src/node.rs:354]: Indeed, that's what I want to do once I everything works with objc2 0.5.

**Comment by DataTriny** [comment]
icrate 0.1.0 has been released, although the path towards reducing the number of features isn't clear at the moment. I currently don't have access to my Mac, I'll come back to this next year anyway.

**Comment by DataTriny** [comment]
I think we use some methods that might not need to be marked unsafe:

- `NSView::convertRect_toView`,
- `NSView::isFlipped`,
- `NsWindow::convertPointFromScreen`,
- ``NSWindow::convertRectToScreen

Otherwise this PR is ready for review.

**Comment by DataTriny** [comment]
The safe alternative to `Id::cast` in this context is `Id::into_super`, but then we have to call it multiple times to go up the inheritance chain.

I think the only place where we have both an `NSObject` and an `AnyObject` is in `PlatformNode::parent`. Since we get an `AnyObject` from `NSView::accessibilityParent`, and since I don't think it is possible to transform it into an `NSObject`, I don't think we have much choice here.

**Comment by mwcampbell** [comment]
I'd rather call `Id::into_super` multiple times to reduce use of `unsafe`.

**Comment by madsmtm** [comment]
> I think we use some methods that might not need to be marked unsafe:

I think I'm considering a way to automatically mark simple things like this as safe, but it's a difficult trade-off between usability and soundness.

I've opened https://github.com/madsmtm/objc2/pull/549 for the specific ones you listed, at least.

> `Id::into_super`, but then we have to call it multiple times to go up the inheritance chain.

Tracking the more generic way in https://github.com/madsmtm/objc2/issues/518.

I agree with calling `into_super` multiple times for now, even though it's quite verbose.

> when to use `NSObject` versus `AnyObject`.

`NSObject` is the root class of almost all classes, but in theory a user could use a subclass of `NSProxy`, or they could define their own root class. So in those cases, where we don't know _anything_ about the object, we use `AnyObject`.

If you _can_ use `NSObject`, then you should. But if a protocol you're implementing takes `&AnyObject`, or something returns `Id<AnyObject>`, then you must use that (and you'll have to live with the limited functionality).

**Comment by mwcampbell** [comment]
@DataTriny I'd like to replace `Id::cast` with `Id::into_super` (making multiple calls as necessary). This PR also needs to be brought up to date with the role description support we just merged. Once those fixes are done, I'm ready to merge this, and we can release everything we have pending, including the Python bindings.

**Comment by DataTriny** [comment]
Yup! Will do this tomorrow though.

**Comment by mwcampbell** [comment]
@DataTriny Would you mind if I do these fixes and then merge this PR? That's one advantage of us both working in the upstream repo.

**Comment by DataTriny** [comment]
Of course, go ahead!

**Pull request merged**